### PR TITLE
add seed cipher util code

### PR DESCRIPTION
**Pull request by neonphog**
DEPENDS ON https://github.com/holochain/lair/pull/46

- adds utilites that make writing the seed_cipher code possible
  - `pw_hash_limits` - argon2id memory and ops limit controls
  - `pw_utils` - password hashing utilities
  - `seed_bundle` - serialization code to translate between ergonomic rust structs and compact msgpack array format
  - `u8array` - fixed-size array newtype with lots of ergonomic from impls

### CODE REVIEW

File: crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
+use super::*;
+
+/// For SeedCipherSecurityQuestions (and the "Locked" struct)
+/// we need to be able to translate three answers into a semi-deterministic
+/// passphrase. This process involves lower-casing, trimming, and concatonating.
```
Comment (maackle) [inline crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs:5]:
```suggestion
/// passphrase. This process involves lower-casing, trimming, and concatenating.
```

File: crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        // this needs a mutable buffer, so we have to do this in out memory
+        out.make_ascii_lowercase();
+    }
+
+    // return the read-only concatonated passphrase
+    Ok(out.to_read())
+}
+
+/// Use the given passphrase to generate a deterministic secret with argon.
+/// Use that secret to secrestream encrypt the given seed.
```
Comment (maackle) [inline crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs:66]:
```suggestion
/// Use that secret to secretstream encrypt the given seed.
```

File: crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
+use super::*;
+
+/// For SeedCipherSecurityQuestions (and the "Locked" struct)
+/// we need to be able to translate three answers into a semi-deterministic
+/// passphrase. This process involves lower-casing, trimming, and concatonating.
+pub(crate) fn process_security_answers<A1, A2, A3>(
```
Comment (maackle) [inline crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs:6]: Are security questions really secure enough for protecting something as important as a seed?

File: crates/hc_seed_bundle/src/seed_cipher/seed_bundle.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+                &question_list.2,
+                header,
+                cipher,
+            )
+                .serialize(serializer),
+        }
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for SeedCipher {
```
Comment (maackle) [inline crates/hc_seed_bundle/src/seed_cipher/seed_bundle.rs:142]: Can you add some roundtrip tests?

File: crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs (hunk 3 above)
Comment (neonphog) [reply crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs:6]: I'd submit it's bitwise probably more data than most people's passwords, and it's optional : )

File: crates/hc_seed_bundle/src/seed_cipher/seed_bundle.rs (hunk 4 above)
Comment (neonphog) [reply crates/hc_seed_bundle/src/seed_cipher/seed_bundle.rs:142]: The tests are all in the final PR https://github.com/holochain/lair/pull/49

**Review by maackle — APPROVED** [review]

**Pull request merged**